lib/msal-common/src/response/ResponseHandler.ts

```diff
@@ -51,6 +54,50 @@ export class ResponseHandler {
             tenantId: idTokenObj.claims.tid,
         };
     }
+
+    private saveToken(originalTokenResponse: TokenResponse, authority: string, serverTokenResponse: ServerAuthorizationTokenResponse, clientInfo: ClientInfo): TokenResponse {
+        const tokenResponse = { ...originalTokenResponse };
+        // Set consented scopes in response
+        const requestScopes = ScopeSet.fromString(serverTokenResponse.scope, this.clientId, false);
+        tokenResponse.scopes = requestScopes.asArray();
+
+        const accessTokenCacheItems = this.cacheManager.getAllAccessTokens(this.clientId, authority);
+
+        for (let i = 0; i < accessTokenCacheItems.length; i++) {
+            const accessTokenCacheItem = accessTokenCacheItems[i];
+            if (accessTokenCacheItem.key.homeAccountIdentifier === tokenResponse.account.homeAccountIdentifier) {
+                const cachedScopes = ScopeSet.fromString(accessTokenCacheItem.key.scopes, this.clientId, false);
+
+                if(cachedScopes.intersectingScopeSets(tokenResponse.scopes)) {
+                    this.cacheStorage.removeItem(JSON.stringify(accessTokenCacheItem.key));
+                }
+            }
+        }
+
+        // Expiration calculation
+        const expiresIn = TimeUtils.parseExpiresIn(serverTokenResponse.expires_in);
+        const expiration = TimeUtils.now() + expiresIn;
+
+        // Save access token in cache
+        const accessTokenKey = new AccessTokenKey(authority, this.clientId, serverTokenResponse.scope, clientInfo.uid, clientInfo.utid, this.cryptoObj);
+        const accessTokenValue = new AccessTokenValue(serverTokenResponse.access_token, serverTokenResponse.id_token, serverTokenResponse.refresh_token, expiresIn.toString(), tokenResponse.account.homeAccountIdentifier);
+        this.cacheStorage.setItem(JSON.stringify(accessTokenKey), JSON.stringify(accessTokenValue));
+
+        // Save tokens in cache
+        tokenResponse.accessToken = serverTokenResponse.access_token;
+        tokenResponse.idToken = serverTokenResponse.id_token;
+        tokenResponse.refreshToken = serverTokenResponse.refresh_token;
+        tokenResponse.expiresOn = new Date(expiration * 1000);
+        return tokenResponse;
```

Comment:
`tokenRespone` doesn't really need to be created until the very end, and I would argue it makes the code harder to follow by creating it (at the top), mutating it, using the updated values, mutating again, then returning.

Recommendation:

```js
const requestScopes = ScopeSet.fromString(serverTokenResponse.scope, this.clientId, false);
const requestScopesArray = requestScopes.asArray();

// ... remove references to `tokenResponse` as noted above

return {
    ...originalTokenResponse,
    scopes: requestScopesArray,
    accessToken: serverTokenResponse.access_token,
    idToken: serverTokenResponse.id_token,
    refreshToken: serverTokenResponse.refresh_token,
    expiresOn: new Date(expirationSeconds * 1000)
}
```

Reply:
+1. Improves readability.

Reply:
I have addressed this here, but will address this more in a future PR, as there is more work to be done here for the silent renewal flow.